src/util/string.h

```diff
@@ -37,3 +45,29 @@ inline QString coloredLinkString(
             QStringLiteral(";\" href=\"") + baseUrl + extUrl +
             QStringLiteral("\">") + text + QStringLiteral("</a>");
 }
+
+// Check if the extension from the file filter was added to the file base name.
+// Otherwise add it manually.
+// Works around https://bugreports.qt.io/browse/QTBUG-27186
+inline QString filePathWithSelectedExtension(const QString& fileLocationInput,
```

Comment:
```suggestion
[[nodiscard]] QString filePathWithSelectedExtension(const QString& fileLocationInput,
```
the compiler usually knows when its best to inline. I'm unsure about the inline keyword here. I still don't understand it tbh. It might make sense / is required in a header file like this.

Reply:
That's not correct imho. If you don't declare definitions of free functions in header files as inline you might end up with multiple definitions (= implementations) and get linker errors due to these conflicts.

C++ legacy cruft.

Reply:
This complex function does not need to be inlined. Please move it into string.cpp.

Reply:
string.h is also the wrong place. This functionality belongs to file paths and file systems and has nothing to do with generic strings.

Reply:
What place do you suggest?

Reply:
Thanks for the clarifications Uwe, so omitting inline results in ODR violations then.

> What place do you suggest?

what about `src/util/file.{h,cpp}`?

Reply:
hmm `src/util/file.cpp` has only a `MDir` class, seems to be for directories only.
Shall I add a new class, like `FilePath`?

Reply:
No need to introduce class if a function can do just as well IMO. Just declare the free function in the header file and then implement it in the cpp.

Reply:
Or add it as a static function to `mixxx::FileInfo` that returns a new instance, i.e. a factory function?

Just a pragmatic proposal to avoid adding new source files with unclear purpose.

Reply:
Can you elaborate? I can't spot any `mixxx::FileInfo` namespace or class in our sourcetree.

Reply:
util/fileinfo.h

Reply:
That is not present on 2.3